File: 04_Transformations/java/Parser.java
/*========== Parser.java ==========

Goes through a file and performs all of the actions listed.
The file follows the following format:
     Every command is a single character that takes up a line
     Any command that requires arguments must have those arguments in the second line.
     The commands are as follows:
         l: add a line to the edge matrix - 
	    takes 6 arguemnts (x0, y0, z0, x1, y1, z1)
	 i: set the transform matrix to the identity matrix - 
	 s: create a scale matrix, 
	    then multiply the transform matrix by the scale matrix - 
	    takes 3 arguments (sx, sy, sz)
	 t: create a translation matrix, 
	    then multiply the transform matrix by the translation matrix - 
	    takes 3 arguments (tx, ty, tz)
	 x: create an x-axis rotation matrix,
	    then multiply the transform matrix by the rotation matrix -
	    takes 1 argument (theta)
	 y: create an y-axis rotation matrix,
	    then multiply the transform matrix by the rotation matrix -
	    takes 1 argument (theta)
	 z: create an z-axis rotation matrix,
	    then multiply the transform matrix by the rotation matrix -
	    takes 1 argument (theta)
	 a: apply the current transformation matrix to the 
	    edge matrix
	 g: draw the lines of the edge matrix to the Frame
	    save the Frame to a file -
	    takes 1 argument (file name)
	 q: end parsing

See the file script for an example of the file format


IMPORTANT MATH NOTE:
the trig functions in java.Math  use radian mesure, but us normal
humans use degrees, so the file will contain degrees for rotations,
be sure to conver those degrees to radians (Math.PI is the constant
for PI)


jdyrlandweaver
=========================*/

import java.io.*;
import java.util.*;
import java.awt.*;

public class Parser {
    
    /*===========================
      transform is the master transform Matrix
      em is the master EdgeMatrix
      f is the frame used for drawing and saving
      =========================*/
    private Matrix transform;
    private EdgeMatrix em;
    private Frame f;

    public Parser()
    {
	f = new Frame();
	transform = new Matrix(4, 4);
	em = new EdgeMatrix();
	transform.ident();
    }


    /*========     public void parseFile() ==========
      Inputs:   BufferedReader in  
      Returns: 

      Goes through the input stream referred to by in,
      scans it for the commands listed above, and performs
      the required commands.
      
      03/08/12 19:20:55
      jdyrlandweaver
      ====================*/
    public void parseFile(BufferedReader in) throws IOException
    {
	String[] args;
	String line;
	for(line = in.readLine(); line != null && !line.equals("q"); line = in.readLine())
	{
	    if(line.length() != 1)
	    {
		continue;
	    }
	    switch(line.charAt(0))
	    {
	    case 'l':
		line = in.readLine();
		args = line.split(" ");
		em.addEdge(Integer.parseInt(args[0]),
			   Integer.parseInt(args[1]),
			   Integer.parseInt(args[2]),
			   Integer.parseInt(args[3]),
			   Integer.parseInt(args[4]),
			   Integer.parseInt(args[5]));
		break;
	    case 'i':
		transform.ident();
		break;
	    case 's':
		line = in.readLine();
		args = line.split(" ");
		Matrix s = new Matrix(4, 4);
		s.makeScale(Double.parseDouble(args[0]),
			    Double.parseDouble(args[1]),
			    Double.parseDouble(args[2]));
		transform.matrixMult(s);
		break;
	    case 't':
		line = in.readLine();
		args = line.split(" ");
		Matrix t = new Matrix(4, 4);
		t.makeTranslate(Double.parseDouble(args[0]),
				Double.parseDouble(args[1]),
				Double.parseDouble(args[2]));
		transform.matrixMult(t);
		break;
	    case 'x':
		line = in.readLine();
		args = line.split(" ");
		Matrix rx = new Matrix(4, 4);
		rx.makeRotX(Double.parseDouble(args[0]) * 3.141592653 / 180.0);
		transform.matrixMult(rx);
		break;
	    case 'y':
		line = in.readLine();
		args = line.split(" ");
		Matrix ry = new Matrix(4, 4);
		ry.makeRotY(Double.parseDouble(args[0]) * 3.141592653 / 180.0);
		transform.matrixMult(ry);
		break;
	    case 'z':
		line = in.readLine();
		args = line.split(" ");
		Matrix rz = new Matrix(4, 4);
		rz.makeRotZ(Double.parseDouble(args[0]) * 3.141592653 / 180.0);
		transform.matrixMult(rz);
		break;
	    case 'a':
		em.matrixMult(transform);
		break;
	    case 'g':
		line = in.readLine();
		args = line.split(" ");
		if(args.length == 3)
		{
		    f = new Frame(Integer.parseInt(args[1]), Integer.parseInt(args[2]));
		}
		else
		{
		    f = new Frame();
		}
		f.drawLines(em, new Color(0, 255, 0));
		f.save(args[0]);
		break;
	    }
	}
    }
}

File: 04_Transformations/c/parser.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <unistd.h>

#include "ml6.h"
#include "display.h"
#include "draw.h"
#include "matrix.h"
#include "parser.h"


/*======== void parse_file () ==========
Inputs:   char * filename 
          struct matrix * transform, 
          struct matrix * pm,
          screen s
Returns: 

Goes through the file named filename and performs all of the actions listed in that file.
The file follows the following format:
     Every command is a single character that takes up a line
     Any command that requires arguments must have those arguments in the second line.
     The commands are as follows:
         l: add a line to the edge matrix - 
	    takes 6 arguemnts (x0, y0, z0, x1, y1, z1)
	 i: set the transform matrix to the identity matrix - 
	 s: create a scale matrix, 
	    then multiply the transform matrix by the scale matrix - 
	    takes 3 arguments (sx, sy, sz)
	 t: create a translation matrix, 
	    then multiply the transform matrix by the translation matrix - 
	    takes 3 arguments (tx, ty, tz)
	 x: create an x-axis rotation matrix,
	    then multiply the transform matrix by the rotation matrix -
	    takes 1 argument (theta)
	 y: create an y-axis rotation matrix,
	    then multiply the transform matrix by the rotation matrix -
	    takes 1 argument (theta)
	 z: create an z-axis rotation matrix,
	    then multiply the transform matrix by the rotation matrix -
	    takes 1 argument (theta)
	 a: apply the current transformation matrix to the 
	    edge matrix
	 v: draw the lines of the edge matrix to the screen
	    display the screen
	 g: draw the lines of the edge matrix to the screen
	    save the screen to a file -
	    takes 1 argument (file name)
	 q: end parsing

See the file script for an example of the file format

IMPORTANT MATH NOTE:
the trig functions int math.h use radian mesure, but us normal
humans use degrees, so the file will contain degrees for rotations,
be sure to conver those degrees to radians (M_PI is the constant
for PI)

03/08/12 16:22:10
jdyrlandweaver
====================*/
void parse_file ( char * filename, 
                  struct matrix * transform, 
                  struct matrix * pm,
                  screen s)
{
  char line[512];
  color c;
  double args[6];
  struct matrix* tmp;
  FILE* file = fopen(filename, "r");
  if(file)
  {
    for(fgets(line, 512, file); strcmp(line, "q\n"); fgets(line, 512, file))
    {
      sleep(2);
      if(!strcmp(line, "i\n"))
      {
	ident(transform);
      }
      else if(!strcmp(line, "a\n"))
      {
	matrix_mult(transform, pm);
      }
      else if(!strcmp(line, "v\n"))
      {
	c.red = 0;
	c.green = 0;
	c.blue = 255;
	clear_screen(s);
	draw_lines(pm, s, c);
	display(s);
      }
      else if(!strcmp(line, "l\n"))
      {
	fgets(line, 512, file);
	sscanf(line, "%lf %lf %lf %lf %lf %lf ", &args[0], &args[1], &args[2], 
	       &args[3], &args[4], &args[5]);
	add_edge(pm, args[0], args[1], args[2], 
	       args[3], args[4], args[5]);
	printf("%lf %lf %lf %lf %lf %lf\n", args[0], args[1], args[2],
               args[3], args[4], args[5]);
      }
      else if(!strcmp(line, "s\n"))
      {
	fgets(line, 512, file);
	sscanf(line, "%lf %lf %lf ", &args[0], &args[1], &args[2]);
	tmp = make_scale(args[0], args[1], args[2]);
	matrix_mult(tmp, transform);
	free_matrix(tmp);
	printf("%lf %lf %lf\n", args[0], args[1], args[2]);
      }
      else if(!strcmp(line, "t\n"))
      {
	fgets(line, 512, file);
	sscanf(line, "%lf %lf %lf", &args[0], &args[1], &args[2]);
	tmp = make_translate(args[0], args[1], args[2]);
	matrix_mult(tmp, transform);
	free_matrix(tmp);
	printf("%lf %lf %lf\n", args[0], args[1], args[2]);
      }
      else if(!strcmp(line, "x\n"))
      {
	fgets(line, 512, file);
	sscanf(line, "%lf ", &args[0]);
	tmp = make_rotX(args[0] * 3.14159265358979323846264338327950 / 180.0);
	matrix_mult(tmp, transform);
	free_matrix(tmp);
	printf("%lf\n", args[0]);
      }
      else if(!strcmp(line, "y\n"))
      {
	fgets(line, 512, file);
	sscanf(line, "%lf ", &args[0]);
	tmp = make_rotY(args[0] * 3.14159265358979323846264338327950 / 180.0);
	matrix_mult(tmp, transform);
	free_matrix(tmp);
	printf("%lf\n", args[0]);
      }
      else if(!strcmp(line, "z\n"))
      {
	fgets(line, 512, file);
	sscanf(line, "%lf", &args[0]);
	tmp = make_rotZ(args[0] * 3.14159265358979323846264338327950 / 180.0);
	matrix_mult(tmp, transform);
	free_matrix(tmp);
	printf("%lf\n", args[0]);
      }
      else if(!strcmp(line, "g\n"))
      {
	fgets(line, 512, file);
	strtok(line, "\n");
	c.red = 0;
	c.green = 0;
	c.blue = 255;
	clear_screen(s);
	draw_lines(pm, s, c);
	save_extension(s, line);
	printf("%s\n", line);
      }
    }
  }
  else
  {
    printf("Invalid filename passed: %s", filename);
  }
  fclose(file);
}
